Unify error handling in direct chain

abailly-iohk opened the pull request:
This PR attempts at simplifying the handling of `postTx`: Posting a transaction may throw an `InvalidTxError` either when constructing the transaction, when trying to balance it or upon a timeout. Then the `Node`'s core event handling loop catches those errors and notify them to the clients.

abailly committed:
Provide Arbitrary instance for InvalidTx errors

abailly committed:
Replace convoluted postTx error handling by exception throwing

abailly committed:
Reinstate timeout when posting transactions

This timeout is needed because the wallet can 'retry' to submit,
waiting for transaction containing marked UTXO to appear on chain. It
cannot be put elsewhere because all other actions are in STM, so it
must appear at the toplevel, eg. in IO.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -156,18 +156,18 @@ withDirectChain tracer networkMagic iocp socketPath keyPair party cardanoKeys ca
+                      )
+                      `catch` \(e :: InvalidTxError CardanoTx) -> do
+                        traceWith tracer $ PostingTxFailed tx e
+                        throwIO e
```
Rather than catching, logging & rethrowing here, I'd expect that to happen where `postTx` is used, e.g. to not have the thread / process crash. Which brings me to the point: What should happen in these cases, should our "HeadLogic" be able to react on this situation? Shall it produce a server output for the clients? I guess that catching and converting it to an `Event` would be necessary for that?

abailly committed:
Document tx posting failure log entries

This lead me to remove the `TxIn` type that was defined in the
`Ledger` module because it lead to complications in the serialisation
process of the errors.

abailly committed:
Remove exception handling code from Direct chain

It is expected that clients to `postTx` will need to handle the
possibility it posts an exception. This will be addressed in a later
PR for the case of the Head protocol.

ghost commented:
@KtorZ @ch1bo Are you ok to merge this?

abailly committed:
Introduced new output message denoting on-chain failure

Right now this is rather unhelpful as there's probably not much a
client can do when faced with these kind of errors, but at least the
node won't crash immediately and we can hope to build some error
handling logic later.

abailly committed:
Add failing test for on-chain error notifications

abailly committed:
Catch InvalidTxError and report them to clients

abailly committed:
Remove unused log message

We don't log posting failures anymore in the DirectChain component but
report them upstream in the Node to clients through ServerOutput message.

abailly committed:
Report on-chain errors in the TUI

ghost commented:
On-chain errors are now reported in the TUI. Note that this is not the case for _all_ errors as there still are a few `error` calls in `DirectChain`.

![Screenshot 2021-12-23 at 13 12 26](https://user-images.githubusercontent.com/79840582/147239000-fca7cbd7-de4b-498d-8ab0-a5ccedefafa3.png)

KtorZ commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -156,18 +156,14 @@ withDirectChain tracer networkMagic iocp socketPath keyPair party cardanoKeys ca
+                            fromPostChainTx wallet (Testnet networkMagic) headState cardanoKeys tx
+                              >>= finalizeTx wallet headState
+                              >>= writeTQueue queue
+                      when (isNothing res) $ throwIO (NoSeedInput @CardanoTx)
```
I don't understand this exception here. It's not clear to me why the consequence of timing out means that there's no seed input? The "NoSeedInput" exception is already thrown within `fromPostChainTx` which seems like the right place? I recall `finalizeTx` is retrying when no payment utxo is found, so perhaps this should be the exception instead? Still, I wouldn't have it at this level, but rather, thrown by the lower levels (i.e. `finalizeTx` itself).

KtorZ commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -179,7 +183,7 @@ processEffect HydraNode{hn, oc, server, eq, env = Environment{party}} tracer e =
     ClientEffect i -> sendOutput server i
     NetworkEffect msg -> broadcast hn msg >> putEvent eq (NetworkEvent msg)
-    OnChainEffect tx -> postTx oc tx
+    OnChainEffect tx -> postTx oc tx `catch` \(ex :: InvalidTxError tx) -> sendOutput server (PostTxOnChainFailed tx ex)
```
This *should* be an event to allow the server to react to it; for now, the only sensible reaction would be sending an output to the client (like here) but later on, have extra error handling logic, retrying etc..

KtorZ commented on hydra-node/test/Hydra/NodeSpec.hs:
```diff
@@ -151,3 +172,18 @@ recordNetwork node = do
+
+  patchedNode ref = node{server = Server{sendOutput = recordMsg ref}}
+
+  queryMsgs = readIORef
```
I guess there;s a good opportunity for refactoring this to make it more DRY with the `recordNetwork` just above.

ch1bo committed:
Move retry into finalizeTx and add a dedicated exception

This limits the hack to the Direct module. Also included a comment on
why this workaround exists.

ch1bo committed:
Rename InvalidTxError -> PostTxError

ch1bo committed:
Add PostTxError event documentation

ch1bo committed:
Handle PostTxError in HeadLogic

The forwarding of this Event as a ClientEffect might seem verbose, but
this pulls in logic "from the edges" into the core logic and keeps
complexity of the "glue code" minimal.

ch1bo committed:
DRY message recorders in NodeSpec

KtorZ reviewed (approved):
Ok.
